Lib/glob.py

```diff
@@ -310,3 +157,229 @@ def translate(pat, *, recursive=False, include_hidden=False, seps=None):
+    """
+    if not parts:
+        return _select_exists
+    part = parts[0]
+    if recursive and part == '**':
+        selector = _recursive_selector
+    elif magic_check.search(part) is not None:
+        selector = _wildcard_selector
+    else:
+        selector = _literal_selector
+    return selector(part, parts[1:], recursive, include_hidden)
+
+
+def _literal_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects a literal descendant of a given path.
+    """
+    is_special = part in _special_parts
+    while parts and magic_check.search(parts[0]) is None:
+        # Consume next non-wildcard component (speeds up joining).
+        if parts[0] not in _special_parts:
+            is_special = False
+        part += os.path.sep + parts[0]
+        parts = parts[1:]
+
+    select_next = _selector(parts, recursive, include_hidden)
+
+    def select_literal(path, rel_path, dir_fd, exists):
+        path = _add_trailing_slash(path) + part
+        rel_path = _add_trailing_slash(rel_path) + part
+        yield from select_next(path, rel_path, dir_fd, exists and is_special)
+    return select_literal
+
+
+def _wildcard_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects direct children of a given path,
+    filtering by pattern.
+    """
+    if include_hidden and part == '*':
+        match = None  # Skip generating a pattern that would match all inputs.
+    else:
+        match = _compile_pattern(part, recursive, include_hidden)
+
+    if parts:
+        select_next = _selector(parts, recursive, include_hidden)
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                if fd is not None:
+                    prefix = _add_trailing_slash(path)
+                with os.scandir(arg) as scandir_it:
+                    entries = list(scandir_it)
+                for entry in entries:
+                    if match is None or match(entry.name):
+                        try:
+                            if entry.is_dir():
+                                entry_path = entry.path
+                                if fd is not None:
+                                    entry_path = prefix + entry_path
+                                yield from select_next(entry_path, entry.name, fd, True)
+                        except OSError:
+                            pass
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    else:
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                prefix = _add_trailing_slash(path)
+                for name in os.listdir(arg):
```

Comment:
This is the one place where I think we probably could use `os.scandir()` and avoid reading the result into memory, because this function specifically handles terminal non-recursive wildcard segments, and so there's no possibility for further recursion.

... but in my testing it's consistently slower, even in directories with thousands of files.

Reply:
That is probably expected, scandir yields more complicated instances rather than a list of strs.

It's possible to plumb through whether or not this was an `iglob` or a `glob` and choose which way to do it, but that feels like unnecessary complexity.

Consider at what point this PR may be over-optimizing if just measuring total time. A potentially more interesting thing to measure is the total number of system calls running test_glob.py before and after this change.  (particularly directory reading and stat syscalls). `strace` is your friend there.

Reply:
Some `strace -c` results, everything above 1%. I measured the whole `test_glob` run so there's a bit of non-glob activity mixed in.

On `main`:

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 29.05    0.020463           1     16182           getdents64
 24.66    0.017368           1      8844       546 openat
 22.89    0.016127           1     14198      1072 newfstatat
 13.81    0.009726           1      8924           close
  2.47    0.001739           0      1881           fcntl
  2.32    0.001632           3       414           unlinkat
  1.38    0.000970           5       185           mkdir
  1.01    0.000714           0       765         3 lseek
```

This branch with `main` merged:

```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 33.02    0.023138           1     16316           getdents64
 25.74    0.018037           2      8461        68 openat
 18.01    0.012619           1     10696       878 newfstatat
 15.36    0.010759           1      9023           close
  2.07    0.001451           7       201           mkdir
  2.07    0.001448           0      1893           fcntl
  1.76    0.001235           2       454           unlinkat
```

I think the `newfstatat` saving is from avoiding unnecessary `isdir()` and `lstat()` calls. The recursive globbing improvement doesn't show up because `test_glob` doesn't create many deep trees.

Reply:
`os.scandir()` is used for optimization in `glob()`. It allows to avoid numerous `os.stat()` calls.

But the iteration of `os.scandir()` should be finished before try to handle other directories. `os.scandir()` consumes a limited resource -- a file descriptor. If you have many non-closed `os.scandir()` iterators, you can easily exceed the limit on a large tree. This is unacceptable. This is why all uses of `os.scandir()` (here, in os.walk, shutil.rmtree, etc) save the entries in  a list before proceeding the subdirectories.

In some cases it is possible to save only directory entries, in assumption that usually the number of files is significantly larger than the number of directories.

Reply:
> It's possible to plumb through whether or not this was an `iglob` or a `glob` and choose which way to do it, but that feels like unnecessary complexity.

I've just tried this, and it works nicely without being particularly complicated. Should we expose it as a keyword argument to `iglob()` and/or make it the default behaviour for `glob()`?

Reply:
I've played around with exposing a new `close_dirs` keyword argument that would default to `True`, but I think it's too hard to explain and use safely. Resolving this thread but feel free to re-open!